src/vos/vos_ts.h

```diff
@@ -506,50 +603,93 @@ vos_ts_rh_update(struct vos_ts_entry *entry, daos_epoch_t read_time,
 	uuid_copy(entry->te_tx_rh, tx_id);
 }
 
-/** Check the read low timestamp at current entry.
+/** Internal API to check read conflict of a given entry */
+bool
+vos_ts_check_read_conflict(struct vos_ts_set *ts_set, int idx,
+			   daos_epoch_t write_time);
+
+/** Checks the set for read/write conflicts
  *
- * \param[in]	ts_set		The timestamp set
- * \param[in]	write_time	The write time
+ * \param[in]	ts_set		The timestamp read set
+ * \param[in]	write_time	The time of the update
  *
- * \return	true	Conflict
- *		false	No conflict (or no timestamp set)
+ * \return	true		Conflict
+ *		false		No conflict (or no timestamp set)
  */
-static inline bool
-vos_ts_check_rl_conflict(struct vos_ts_set *ts_set, daos_epoch_t write_time)
+static inline int
+vos_ts_set_check_conflict(struct vos_ts_set *ts_set, daos_epoch_t write_time)
 {
-	struct vos_ts_entry	*entry;
+	int			 i;
 
-	entry = vos_ts_set_get_entry(ts_set);
-	if (entry == NULL || write_time > entry->te_ts_rl)
+	if (ts_set == NULL)
 		return false;
 
-	if (write_time != entry->te_ts_rl)
-		return true;
+	if ((ts_set->ts_cflags & VOS_TS_WRITE_MASK) == 0)
+		return false;
 
-	return uuid_compare(ts_set->ts_tx_id, entry->te_tx_rl) != 0;
+	for (i = 0; i < ts_set->ts_init_count; i++) {
+		/** Will check the appropriate read timestamp based on the type
+		 *  of the entry at index i.
+		 */
+		if (vos_ts_check_read_conflict(ts_set, i, write_time))
+			return true;
+	}
+
+	return false;
 }
 
-/** Check the read high timestamp at current entry.
+/** Update the read timestamps for the set after a successful operation
  *
- * \param[in]	ts_set		The timestamp set
- * \param[in]	write_time	The write time
- *
- * \return	true	Conflict
- *		false	No conflict (or no timestamp set)
+ *  \param[in]	ts_set		The timestamp set
+ *  \param[in]	read_time	The new read timestamp
  */
-static inline bool
-vos_ts_check_rh_conflict(struct vos_ts_set *ts_set, daos_epoch_t write_time)
+static inline void
+vos_ts_set_update(struct vos_ts_set *ts_set, daos_epoch_t read_time)
 {
-	struct vos_ts_entry	*entry;
+	struct vos_ts_set_entry	*se;
+	uint32_t		 high_mask = 0;
+	uint32_t		 low_mask = 0;
+	int			 i;
 
-	entry = vos_ts_set_get_entry(ts_set);
-	if (entry == NULL || write_time > entry->te_ts_rh)
-		return false;
+	if (ts_set == NULL)
+		return;
 
-	if (write_time != entry->te_ts_rh)
-		return true;
+	if ((ts_set->ts_cflags & VOS_TS_READ_MASK) == 0)
+		return;
+
+	for (i = 0; i < ts_set->ts_init_count; i++) {
+		se = &ts_set->ts_entries[i];
+
+		switch (se->se_etype) {
+		case VOS_TS_TYPE_CONT:
+			high_mask = VOS_TS_READ_MASK;
+			low_mask = VOS_TS_READ_CONT;
+			break;
+		case VOS_TS_TYPE_OBJ:
+			high_mask = VOS_TS_READ_OBJ_CHILD;
+			low_mask = VOS_TS_READ_OBJ;
+			if (ts_set->ts_max_type > VOS_TS_TYPE_OBJ)
+				break;
+		case VOS_TS_TYPE_DKEY:
+			high_mask |= VOS_TS_READ_DKEY_CHILD;
```

Comment:
The fall-through and the _CHILD are redundant, are they not? Just want to make sure it's intentional.

Reply:
yeah, they are but if we go straight to DKEY, it needs to be set